ASS/AWS.py

```diff
@@ -122,13 +141,61 @@ def remove_bucket(self, bucket_name):
-            self.logger.info(f"Start deletion of bucket {bucket}")
+            self.logger.info(f"Start deletion of bucket {bucket_name}")
             bucket.delete()
-            self.logger.info(f"Finished deletion of bucket {bucket}")
+            self.logger.info(f"Finished deletion of bucket {bucket_name}")
         except Exception:
             self.logger.error(f"An error occurred while deleting bucket {bucket_name}")
             raise
 
+    def backup_bucket(self, origin_bucket_name, backup_bucket_name):
+        try:
+            self.logger.info(f"Connect to bucket {origin_bucket_name}")
+            s3 = boto3.client('s3')
+            s3_resource = boto3.resource('s3')
+            self.logger.info(f"Start backup of all objects in bucket {origin_bucket_name}")
+            for key in s3.list_objects(Bucket=origin_bucket_name)['Contents']:
+                file = key['Key']
+                copy_source = {'Bucket': origin_bucket_name, 'Key': file}
+                s3_resource.meta.client.copy(copy_source, backup_bucket_name, f"{origin_bucket_name}/{key['Key']}")
+            self.logger.info(f"Finished backup of bucket {origin_bucket_name} to {backup_bucket_name}")
+        except Exception:
+            self.logger.error(f"An error occurred while taking a backup of bucket {origin_bucket_name}")
+            raise
 
+    def restore_bucket(self, bucket_name):
+        try:
+            origin_bucket_name = f"aws-ass-{ self.get_account_id() }-s3-backup"
+            self.logger.info(f"Connect to bucket {origin_bucket_name}")
+            s3 = boto3.client('s3')
+            s3_resource = boto3.resource('s3')
+            self.logger.info(f"Start restore of all objects in bucket {origin_bucket_name}")
+            paginator = s3.get_paginator('list_objects_v2')
+            pages = paginator.paginate(Bucket=origin_bucket_name, Prefix=f'{bucket_name}/')
+            for page in pages:
+                for obj in page['Contents']:
+                    # full path (e.g. bucket/folder/test.png)
+                    origin_file_key = obj['Key']
+                    # first folder in path (e.g. bucket)
+                    bucket = "/".join(origin_file_key.strip("/").split('/')[0:1])
+                    # ACL tagging
+                    acl = ""
+                    for tag in s3.get_bucket_tagging(Bucket=f"{bucket}")['TagSet']:
+                        if tag['Key'] == "ass:s3:backup-and-empty-bucket-on-stop-acl":
+                            bucket_tag = tag['Value']
+                            acl = {'ACL': f"{bucket_tag}"}
+                        else:
+                            acl = {'ACL': "private"}
+                    # file name (e.g. test.png)
+                    base_name = os.path.basename(origin_file_key)
+                    # path (e.g. folder/test.png)
+                    fn_new_bucket = "/".join(origin_file_key.strip("/").split('/')[1:])
+                    if base_name != '':
```

Comment:
Better to test on the last character of the string, that's simpler and more self explanatory.

```
if not origin_file_key.endswith("/"):
```